Develop

GritsED opened the pull request:
Привет,
Реализовали фичи по ТЗ

GritsED committed:
Добавить удаление фильмов и пользователей

SamuelichSam committed:
Функциональность «Отзывы» fixed after review

SamuelichSam committed:
Функциональность «Отзывы» fixed after review

GritsED committed:
Добавить функциональность "Поиск"

GritsED committed:
Добавить функциональность "Поиск"

SamuelichSam committed:
Функциональность «Общие фильмы»

SamuelichSam committed:
Merge branch 'develop' into add-common-films

SamuelichSam committed:
Функциональность «Общие фильмы» fix chekstyle

SamuelichSam committed:
Merge remote-tracking branch 'origin/add-common-films' into add-common-films

SamuelichSam committed:
Функциональность «Общие фильмы» fixed after review

GritsED committed:
Merge remote-tracking branch 'origin/develop' into develop

ldspr0 committed:
Добавлен функционал по Events (#7)

* Добавлен объект Event, добавлены сервисные классы, Добавлен users/{id}/feed Эндпоинт, обновлена схема дб, добавлено создание евентов для лайков и друзей

ldspr0 committed:
Функциональность «Рекомендации» (#12)

Функциональность «Рекомендации»

GritsED committed:
Merge remote-tracking branch 'origin/add-search' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java

GritsED committed:
Merge branch 'refs/heads/develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java

yuto-jj committed:
Добавление режиссёров в фильмы (#14)

* Добавление режиссёров в фильмы

* Fixed few issues with Directors to pass Postman

GritsED committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java

GritsED committed:
Optimize film queries: removed extra joins, updated methods, and added duplicate director check

GritsED committed:
Вывод самых популярных фильмов по жанру и годам (#13)

* Вывод самых популярных фильмов по жанру и годам

GritsED committed:
Merge branch 'develop' into add-search

GritsED committed:
Fix method searchFilm

GritsED committed:
Добавить функциональность "Поиск" (#10)

* Добавить функциональность "Поиск"

GritsED committed:
Fix code to pass partial Postman tests

Commit:
Fixed issues with Events, TopFilms and Reviews according to Postman requirements.

Commit:
Merge

GritsED committed:
Refactor code: removed unnecessary joins and constructors, added @RequiredArgsConstructor annotation and @Primary annotation

vlolad commented on src/main/java/ru/yandex/practicum/filmorate/dao/ReviewDbStorage.java:
```diff
@@ -0,0 +1,184 @@
+    @Override
+    public void deleteReview(Long id) {
+        log.debug("Received request to remove review with ID: {}", id);
+        //jdbc.update(DELETE_USEFUL, id);
```
Лишний комментарий

vlolad reviewed (changes requested):
Привет!

Работа отличная, не могу перестать хвалить за отлично написанные SQL-запросы. В этой части сплошной зелёный комментарий, как раньше было)

Есть пару моментов по улучшениям и одна критичная правка: при сохранении фильмов операции по вставке в промежуточные таблицы (жанры и режиссеры) выполняются в цикле forEach. Для вставки нескольких значений лучше использовать другой метод, который предполагает пакетную вставку в БД. Так сокращаем число операций и в целом работать приложение будет стабильнее под нагрузкой.

Поздравляю всю команду с отличной работой, доработок совсем немного)
Хороших выходных!

GritsED commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDbStorage.java:
```diff
@@ -2,28 +2,68 @@
+            LEFT JOIN likes l ON f.id = l.film_id
+            LEFT JOIN filmGenre fg ON fg.film_id = f.id
+            LEFT JOIN genres g ON g.id = fg.genre_id
+            WHERE f.id IN (
```
Изначально запрос был такой 
SELECT f.*, m.id AS mpa_id, m.rate, COUNT(l.user_id) AS likes_count, g.id genre
            FROM films f
            JOIN mpa m ON f.mpa_id = m.id
            LEFT JOIN likes l ON f.id = l.film_id
            LEFT JOIN filmGenre fg ON fg.film_id = f.id
            LEFT JOIN genres g ON g.id = fg.genre_id
            JOIN likes l1 ON f.id = l1.film_id
            JOIN likes l2 ON f.id = l2.film_id
            WHERE l1.user_id = ? AND l2.user_id = ?
            GROUP BY f.id, m.id, m.rate, g.id
            ORDER BY likes_count DESC
            
Но у нас в команде есть разраб действующий и он порекомендовал использовать подзапрос. 
© "множественный джоин таблиц (которые могут стать большими), не очень эффективен, на сколько я знаю"
Какой вариант будет лучше, множественный джойн таблиц или подзапрос?

GritsED committed:
Fixed after review

DirectorController, FilmController, UserController:
Remove redundant null checks for @PathVariable parameters

FilmDbStorage, FilmDirectorDbStorage, FilmGenreDbStorage:
Use batchUpdate for improved efficiency

Remove unused storage classes

vlolad commented on src/main/java/ru/yandex/practicum/filmorate/dao/FilmDirectorDbStorage.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public void addDirectorToFilm(Long filmId, List<Director> directorId) {
+        log.debug("Received request to add {} directors to film with ID {}", directorId.size(), filmId);
+        jdbc.batchUpdate(INSERT_FILM_DIRECTOR_QUERY, new BatchPreparedStatementSetter() {
```
Отлично

vlolad reviewed (approved):
Привет!

Всё отлично, недочёты исправлены, всё принято (￣^￣)ゞ
Самое главное, что у вас очень удачно сложилась командная работа (по крайней мере, судя по итоговому результату с:). Не было дублирования кода в разных сервисах, приложение целостное, с выдержанным стилем так сказать. Это бывает куда важнее навыка кодинга

Ну и я не могу не похвалить всех, кто SQL-запросы писал в работе, потому что давно уже не видел такой хорошей работы с БД. Обычно на групповой проект меньше 20 замечаний я не оставляю, так что все молодцы с:

С чистой совестью идём отдыхать, всем хороших остатков выходных!

GritsED committed:
Add project description for group project features
